transport/p2p/src/behavior/discovery.rs

```diff
@@ -0,0 +1,187 @@
+        _local_addr: &libp2p::Multiaddr,
+        _remote_addr: &libp2p::Multiaddr,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        if self.allowed_peers.contains(&peer) {
+            Ok(Self::ConnectionHandler {})
+        } else {
+            Err(libp2p::swarm::ConnectionDenied::new(crate::errors::P2PError::Logic(
+                format!("Connection from '{peer}' is not allowed"),
+            )))
+        }
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        peer: libp2p::PeerId,
+        _addr: &libp2p::Multiaddr,
+        _role_override: libp2p::core::Endpoint,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        if self.allowed_peers.contains(&peer) {
+            Ok(Self::ConnectionHandler {})
+        } else {
+            Err(libp2p::swarm::ConnectionDenied::new(crate::errors::P2PError::Logic(
+                format!("Connection to '{peer}' is not allowed"),
+            )))
+        }
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            libp2p::swarm::FromSwarm::ConnectionEstablished(data) => {
+                *self.connected_peers.entry(data.peer_id).or_insert(0) += 1
+            }
+            libp2p::swarm::FromSwarm::ConnectionClosed(data) => {
+                let v = self.connected_peers.entry(data.peer_id).or_insert(0);
+                if *v > 0 {
+                    *v -= 1;
+                };
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: libp2p::PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+        // Nothing is necessary here, because no ConnectionHandler events should be generated
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p::swarm::ToSwarm<Self::ToSwarm, libp2p::swarm::THandlerInEvent<Self>>> {
+        if let Some(value) = self.pending_events.pop_front() {
+            return std::task::Poll::Ready(value);
+        };
+
+        let poll_result = self.events.poll_next_unpin(cx).map(|e| match e {
+            Some(DiscoveryInput::NetworkUpdate(event)) => match event {
+                NetworkTriggeredEvent::CloseConnection(peer) => {
+                    debug!(peer = %peer, "p2p - discovery - Closing connection (reason: low ping connection quality");
+                    if self.is_peer_connected(&peer) {
+                        self.pending_events.push_back(ToSwarm::CloseConnection {
+                            peer_id: peer,
+                            connection: CloseConnection::default(),
+                        });
+                    }
+                }
+                NetworkTriggeredEvent::UpdateQuality(_, _) => {}
```

Comment:
What happens if the node does a manual active ping to other nodes? Does it contribute to the update of network quality?

Reply:
That is never actually called and is a remnant of the old implementation. Once the `core-network` crate is cleaned up, this functionality will disappear. The behavior itself is responsible for updating the quality in the network db.